src/Lean/Meta/Tactic/Simp/BuiltinSimprocs/Nat.lean

```diff
@@ -68,4 +65,264 @@ builtin_dsimproc [seval] isValue ((OfNat.ofNat _ : Nat)) := fun e => do
   let_expr OfNat.ofNat _ _ _ ← e | return .continue
   return .done e
 
+/-- A literal natural number or a base + offset expression. -/
+private inductive NatOffset where
+  /- denotes expression definition equal to `n` -/
+  | const (n : Nat)
+  /-- denotes `e + o` where `o` is expression definitionally equal to `n` -/
+  | offset (e o : Expr) (n : Nat)
+
+/- Attempt to parse a `NatOffset` from an expression-/
+private partial def NatOffset.asOffset (e : Expr) : Meta.SimpM (Option (Expr × Nat)) := do
+  if e.isAppOfArity ``HAdd.hAdd 6 then
+    let inst := e.appFn!.appFn!.appArg!
+    unless inst.isAppOfArity ``instHAdd 2 do return none
+    unless inst.appArg!.isAppOfArity ``instAddNat 0 do return none
```

Comment:
This is going to fail if the instance is actually `AddMonoid.toAdd` from mathlib, which seems very likely to occur.

Reply:
A small repro:
```lean
import Mathlib.Data.Nat.Basic

example : 1 + 2 = 3 := by
  rewrite [add_comm]
  -- goal is no longer about `instAddNat`
  sorry
  ```

Reply:
Could you make an example showing this?  Using `import Mathlib` is fine.

Reply:
@joehendrix: I just went back in time and posted one :)

Reply:
Note that this was fixed in mathlib in https://github.com/leanprover-community/mathlib/pull/18129

Reply:
For reference, the pattern of matching operators by name and typeclasses by defeq is called "keyed matching". (I learned this name from [Reflexive tactics for algebra, revisited by Kazuhiko Sakaguchi](https://arxiv.org/abs/2202.04330), Section 3.2.)